addons/account/models/account_account.py

```diff
@@ -614,21 +621,28 @@ def _adapt_parent_account_group(self):
         """
         if not self:
             return
-        self.env['account.group'].flush()
+        self.env['account.group'].flush(self.env['account.group']._fields)
         query = """
-            UPDATE account_group agroup SET parent_id = (
-                SELECT parent.id FROM account_group parent
-                WHERE char_length(parent.code_prefix_start) < char_length(agroup.code_prefix_start)
-                AND parent.code_prefix_start <= LEFT(agroup.code_prefix_start, char_length(parent.code_prefix_start))
-                AND parent.code_prefix_end >= LEFT(agroup.code_prefix_end, char_length(parent.code_prefix_end))
-                AND parent.id != agroup.id
-                AND parent.company_id = %(company_id)s
-                ORDER BY char_length(parent.code_prefix_start) DESC LIMIT 1
-            ) WHERE agroup.company_id = %(company_id)s;
+            WITH relation AS (
+       SELECT DISTINCT FIRST_VALUE(parent.id) OVER (PARTITION BY child.id ORDER BY char_length(parent.code_prefix_start) DESC) AS parent_id,
```

Comment:
same question about char_length uniqueness.
With records:
```sql
id | code_prefix_start | code_prefix_end
-----------------------------------------------------
 1 | aa                | zz
 2 | ab                | zz
 3 | aba               | zz
```
record 3 can have 1 or 2 as his parent because ORDER BY doesn't seem to be precise enough.

Reply:
There are constraints that should prevent that
https://github.com/odoo/odoo/blob/f9aa4ab2482b98b074fcbb6a2abb51fd270151e8/addons/account/models/account_account.py#L498-L504

And 
https://github.com/odoo/odoo/blob/f9aa4ab2482b98b074fcbb6a2abb51fd270151e8/addons/account/models/account_account.py#L536-L555

Reply:
There is also this solution
The timing are similar on my tests, but que query plan is a bit different
```sql
            WITH relation AS (
       SELECT DISTINCT ON (child.id)                                                                                                        
                       parent.id AS parent_id,
                       child.id AS child_id
                  FROM account_group parent
                  JOIN account_group child                                                         
                    ON char_length(parent.code_prefix_start) < char_length(child.code_prefix_start)
                   AND parent.code_prefix_start <= LEFT(child.code_prefix_start, char_length(parent.code_prefix_start))
                   AND parent.code_prefix_end >= LEFT(child.code_prefix_end, char_length(parent.code_prefix_end))
                   AND parent.id != child.id
                   AND parent.company_id = child.company_id
              GROUP BY parent.id, child.id
              ORDER BY child.id, char_length(parent.code_prefix_start) DESC
            )              
            UPDATE account_group child
               SET parent_id = relation.parent_id
              FROM relation
             WHERE child.id = relation.child_id;
```